mne/preprocessing/annotate_amplitude.py

```diff
@@ -0,0 +1,235 @@
+# Author: Mathieu Scheltienne <mathieu.scheltienne@fcbg.ch>
+#
+# License: BSD-3-Clause
+
+import numpy as np
+
+from ..fixes import jit
+from ..io import BaseRaw
+from ..annotations import (Annotations, _adjust_onset_meas_date,
+                           _annotations_starts_stops)
+from ..io.pick import _picks_to_idx, _picks_by_type, _get_channel_types
+from ..utils import _validate_type, verbose, logger, _mask_to_onsets_offsets
+
+
+@verbose
+def annotate_amplitude(raw, peak=None, flat=None, bad_percent=5,
+                       min_duration=0.005, picks=None, *, verbose=None):
+    """Annotate raw data based on peak-to-peak amplitude.
+
+    Creates annotations ``BAD_peak`` or ``BAD_flat`` for spans of data where
+    consecutive samples exceed the threshold in ``peak`` or fall below the
+    threshold in ``flat`` for more than ``min_duration``.
```

Comment:
the docstring is much improved!  This sentence, however, has made me realize something that now I'm scratching my head about.  Am I right that the `min_duration` is useful only for `flat` spans?  E.g., for `peak` wouldn't you want something like `max_duration`, where *fast* large changes of signal value are annotated, but *slow* large changes are OK?  If that's correct, then maybe:

1. (to avoid parameter sprawl) we should consider making `min_duration` into something like `window` or `window_length`, and explain that it behaves differently when detecting `peak` vs `flat`; or
2. we should have `peak_window` and `flat_window` (or similar) to let users specify the *min acceptable duration* for large signal changes, and the *max acceptable duration* for flat spans (respectively)

WDYT? Am I misunderstanding something here?

Reply:
I think you are misunderstanding. To annotate either with `BAD_peak` or `BAD_flat`, you need at least a certain number of consecutive points that exceed or fall short of the threshold. For flat, well it's clear that you need at least e.g. 10 consecutive samples to be close to 0. For peak, you need at least e.g. 10 consecutive samples with a difference between each of them above the threshold. 

What does this implies, what will it detect? 
- Drifts-like, large spikes: if the spike is punctual in time, it will detect the spike; if it goes up, hold, and then down, the edges will be detected.
- Signal with abnormally large amplitude: signal that go up and down at a rate larger than the threshold, probably on many small data spans (annotated) that will cover more than `x%` leading to the channel being marked as bad.

So there is no windowing applied to detect large PTP 'windows', only a look upon consecutive samples.

---------------

Overall I'm against (1) since `min_duration` behaves exactly the same for both; and I'm against (2) (with e.g. `min_duration_flat` and `min_duration_peak` to avoid parameter sprawl. If you can't live with a common `min_duration`, then you can call this function 2 times with `peak=None, flat=xxx` and `peak=xxx, flat=None`.

Reply:
> For peak, you need at least e.g. 10 consecutive samples with a difference between each of them above the threshold.

this is indeed what I thought the function is doing, but what I'm suggesting is that it might not align with what users will want or expect.  It may do well for things like muscle artifacts, but DC shifts / flux jumps will likely get missed, because they involve a large change over the course of just a few samples, followed by a comparatively steady/normal signal for a while around the "new baseline", then later another fast jump back down to the old baseline.

Reply:
Exactly the _goes up, hold, and goes down_ case. Yes, this method is not perfect, but has the advantage to be common with `flat` and to give clear fix starts/stop for annotations. 

For the `peak` case, we could do a second sweep of the data with a sliding windows, checking for PTP in the windows, and using the edges of the windows as start/stop for the annotation. 
But then we are starting to populate this function with a lot of different arguments, it requires at least a `window` argument to define the window size and a `peak_window` to define the threshold (could be you want to apply sliding window and not diff, so it has to be different from `peak`).

Reply:
+ also by how much you are sliding the window.. I don't really like this as (1) it will annotate as bads segments that are not actually bad, but that are preceding or following a bad temporal event; and (2) as this is what `Epochs` and its `reject` argument are for.

Reply:
OK, we seem to be on the same page as to what cases this function will/won't catch.  I agree with the proposal that if users want different min_durations for `peak` and `flat` they can just run the function twice. And I'm not too keen on parameter sprawl / making this too complicated...  but I want to make sure that the docstring doesn't mislead anyone into thinking that this function will catch things like DC shifts, and then they wrongly assume there aren't any when the function returns zero annotations.  Maybe I just wasn't paying attention, and saying "consecutive samples" is enough here... but my instinct is that many people will miss the implication of "consecutive samples" and then go on to be surprised when this doesn't work the same way things work when doing epoch rejection.

Reply:
I agree, maybe we could add a small text in the Note section to detail what this consecutiveness implies.

Reply:
Have a look at the note 😉